libtock/crt0.c

```diff
@@ -228,37 +238,82 @@ void _start(void* app_start __attribute__((unused)),
+void _c_start(uint32_t app_start, uint32_t mem_start) {
   struct hdr* myhdr = (struct hdr*)app_start;
 
-  // fix up GOT
-  volatile uint32_t* got_start     = (uint32_t*)(myhdr->got_start + stacktop);
-  volatile uint32_t* got_sym_start = (uint32_t*)(myhdr->got_sym_start + (uint32_t)app_start);
+  // Fix up the Global Offset Table (GOT).
+
+  // Get the address in memory of where the table should go.
+  volatile uint32_t* got_start = (uint32_t*)(myhdr->got_start + mem_start);
+  // Get the address in flash of where the table currently is.
+  volatile uint32_t* got_sym_start = (uint32_t*)(myhdr->got_sym_start + app_start);
+  // Iterate all entries in the table and correct the addresses.
   for (uint32_t i = 0; i < (myhdr->got_size / (uint32_t)sizeof(uint32_t)); i++) {
+    // Use the sentinel here. If the most significant bit is 0, then we know
+    // this offset is pointing to an address in memory. If the MSB is 1, then
+    // the offset refers to a value in flash.
     if ((got_sym_start[i] & 0x80000000) == 0) {
-      got_start[i] = got_sym_start[i] + stacktop;
+      // This is an address for something in memory, and we need to correct the
+      // address now that we know where this app is actually running in memory.
+      // This equation is really:
+      //
+      //     got_entry = (got_stored_entry - original_RAM_start_address) + actual_RAM_start_address
+      //
+      // However, we compiled the app where `original_RAM_start_address` is 0x0,
+      // so we can omit that.
+      got_start[i] = got_sym_start[i] + mem_start;
     } else {
-      got_start[i] = (got_sym_start[i] ^ 0x80000000) + (uint32_t)app_start;
+      // Otherwise, this address refers to something in flash. Now that we know
+      // where the app has actually been loaded, we can reference from the
+      // actual `app_start` address. We also have to remove our fake flash
+      // address sentinel (by ORing with 0x80000000).
+      got_start[i] = (got_sym_start[i] ^ 0x80000000) + app_start;
     }
   }
 
-  // load data section
-  void* data_start     = (void*)(myhdr->data_start + stacktop);
-  void* data_sym_start = (void*)(myhdr->data_sym_start + (uint32_t)app_start);
+  // Load the data section from flash into RAM. We use the offsets from our
+  // crt0 header so we know where this starts and where it should go.
+  void* data_start     = (void*)(myhdr->data_start + mem_start);
+  void* data_sym_start = (void*)(myhdr->data_sym_start + app_start);
   memcpy(data_start, data_sym_start, myhdr->data_size);
 
-  // zero BSS
-  char* bss_start = (char*)(myhdr->bss_start + stacktop);
+  // Zero BSS segment. Again, we know where this should be in the process RAM
+  // based on the crt0 header.
+  char* bss_start = (char*)(myhdr->bss_start + mem_start);
   memset(bss_start, 0, myhdr->bss_size);
 
+  // Do relative data address fixups. We know these entries are stored at the end
+  // of flash and can be located using the crt0 header.
```

Comment:
`rd->len` is hopefully also zero for non-PIC apps?

Reply:
So far, yes.